src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs

```diff
@@ -238,18 +238,21 @@ static void AddAll(HashSet<Checksum> checksums, ChecksumCollection checksumColle
             if (missingChecksumsCount > 0)
             {
                 var missingChecksumsMemory = new ReadOnlyMemory<Checksum>(missingChecksums, 0, missingChecksumsCount);
-                var missingAssets = await RequestAssetsAsync<T>(assetPath, missingChecksumsMemory, cancellationToken).ConfigureAwait(false);
 
-                Contract.ThrowIfTrue(missingChecksumsMemory.Length != missingAssets.Length);
-
-                for (int i = 0, n = missingAssets.Length; i < n; i++)
+                var currentIndex = 0;
+                await RequestAssetsAsync(assetPath, missingChecksumsMemory, (int index, T missingAsset) =>
                 {
-                    var missingChecksum = missingChecksums[i];
-                    var missingAsset = missingAssets[i];
+                    Contract.ThrowIfTrue(currentIndex != index);
+
+                    var missingChecksum = missingChecksums[index];
 
                     AddResult(missingChecksum, missingAsset);
```

Comment:
there are several captures in this lambda.  but they will go away in followup pr.